.circleci/config.yml

```diff
@@ -1,53 +1,66 @@
-version: 2
+version: 2.1
 
-#
 # Reusable Snippets!
 #
 # These are re-used by the various tests below, to avoid repetition.
 #
-run_install_desired_npm: &run_install_desired_npm
-  run:
-    # Due to a bug, npm upgrades from the version of npm that ships with
-    # Node.js 6 (npm v3.10.10) go poorly and generally causes other problems
-    # with the environment.  Since yarn is already available here we can just
-    # use that to work-around the issue.  It's possible that npm cleanup might
-    # prevent this from being necessary, but this installation can be switched
-    # to use `npm` (rather than `yarn`) once Node 6 is no longer tested below.
-    name: Install npm@6, but with yarn.
-    command: sudo yarn global add npm@6
+commands:
+  install_desired_npm:
+    description: "Install the same consistent version of npm"
+    steps:
+      - run:
+          # Due to a bug, npm upgrades from the version of npm that ships with
+          # Node.js 6 (npm v3.10.10) go poorly and generally causes other problems
+          # with the environment.  Since yarn is already available here we can just
+          # use that to work-around the issue.  It's possible that npm cleanup might
+          # prevent this from being necessary, but this installation can be switched
+          # to use `npm` (rather than `yarn`) once Node 6 is no longer tested below.
+          name: Install npm@6, but with yarn.
+          command: sudo yarn global add npm@6
+
+  # These are the steps used for each version of Node which we're testing
+  # against.  Thanks to YAMLs inability to merge arrays (though it is able
+  # to merge objects), every version of Node must use the exact same steps,
+  # or these steps would need to be repeated in a version of Node that needs
+  # something different.  Probably best to avoid that, out of principle, though.
+  common_test_steps:
+    description: "Commands to run on every Node.js environment"
+    steps:
+      - install_desired_npm
+      - checkout
+      - restore_cache:
+          keys:
+            # When lock file changes, use increasingly general patterns to restore cache
+            - npm-v2-{{ .Branch }}-{{ checksum "package-lock.json" }}
+            - npm-v2-{{ .Branch }}-
+            - npm-v2-
+      - run: npm --version
+      - run: npm ci
+      - save_cache:
+          key: npm-v2-{{ .Branch }}-{{ checksum "package-lock.json" }}
+          paths:
+            # This should cache the npm cache instead of node_modules, which is needed because
+            # npm ci actually removes node_modules before installing to guarantee a clean slate.
+            - ~/.npm
```

Comment:
Everything from this line up until `restore_cache` on line 31 is a great candidate for a shared command that could DRY up several of our repositories!